lib/ddtrace/logger.rb

```diff
@@ -10,30 +11,89 @@ class Logger < ::Logger
     def initialize(*args, &block)
       super
       self.progname = LOG_PREFIX
+
+      # Rate limit buckets
+      @buckets = {}
     end
 
     def add(severity, message = nil, progname = nil, &block)
       where = ''
+      c = nil
+      skipped_messages = ''
+
+      # Logging rate of 0 means log everything, so skip bucket checking
+      if Datadog.configuration.logging_rate > 0
+        # Get the call stack to determine who is logging
+        c = caller
+
+        # Get the rate limit bucket key
+        key = rate_limit_key(severity, progname, c)
```

Comment:
- How big can this list of buckets grow?
 - What variables are most likely to drive its growth?
 - Do we need to worry about the number of buckets growing too big?
 - If someone's app was dynamically creating new classes, would that impact how the key is generated?

Reply:
this logger should only be used internally, so the goal is, the list of buckets is basically the total number of unique log messages we emit in code. most people won't have all of them available... unless they use every integration we offer in a single app, but shouldn't be too bad even under that case.

Reply:
> the list of buckets is basically the total number of unique log messages we emit in code

What if that log message contains some kind of variable in it: e.g. "Could not trace Sidekiq Job ID 8427509" or something like that?

Reply:
That is why we want to do it based on the file/line (`caller`-esque) info if we can. We don't want to do it based on the actual message because of the variability.

For your example, we only want one of those errors (regardless of the job id) per 60 seconds.

Reply:
Yup, makes sense, just clarifying.

On another note, how does size of this hash change over time? Does it grow infinitely? How is it pruned?